docs/src/content/docs/book/assembly-functions.mdx

```diff
@@ -0,0 +1,636 @@
+
+:::
+
+Therefore, to access details such as the amount of [nanoToncoins](/book/integers#nanotoncoin) in a message or the [`Address{:tact}`][p] of the sender it's strongly recommended to call the [`context(){:tact}`](/ref/core-common#context) or [`sender(){:tact}`](/ref/core-common#sender) functions instead of attempting to look for those values on the stack.
+
+## Debugging {#debugging}
+
+The number of values the stack has at any given time is called the _depth_, and it's accessible via the [`DEPTH`](https://docs.ton.org/v3/documentation/tvm/instructions#68) instruction. It's quite handy for seeing the number of values before and after calling the assembly functions you're debugging, and can be used within asm logic.
+
+```tact
+asm fun depth(): Int { DEPTH }
+```
+
+To see both the stack depth and the values on it, there's a function in the Core library of Tact: [`dumpStack(){:tact}`](/ref/core-debug#dumpstack). It's great for keeping track of the stack while debugging, although it's computationally expensive and only prints values, not returns them, so use it sparingly and only when testing.
+
+Read more about debugging Tact contracts on the dedicated page: [Debugging](/book/debug).
+
+## Attributes {#attributes}
+
+The following attributes can be specified:
+
+* `inline{:tact}` — does nothing, since assembly functions are always inlined.
+* [`extends{:tact}`](/book/functions#extension-function) — makes it an [extension function](/book/functions#extension-function).
+* [`mutates{:tact}`](/book/functions#mutation-functions) (along with [`extends{:tact}`](/book/functions#extension-function)) — makes it an [extension mutation function](/book/functions#mutation-functions).
+
+Those attributes _cannot_ be specified:
+
+* `abstract{:tact}` — assembly functions must have a body defined.
+* `virtual{:tact}` and `override{:tact}` — assembly functions cannot be defined within a contract or a trait.
+* [`get{:tact}`](/book/functions#getter-functions) — assembly functions cannot be [getters](/book/functions#getter-functions).
+
+```tact
+/// `Builder.storeCoins()` extension function
+asm extends fun storeCoins(self: Builder, value: Int): Builder {
+    STVARUINT16
+}
+
+/// `Slice.skipBits()` extension mutation function
+asm extends mutates fun skipBits(self: Slice, l: Int) {
+    SDSKIPFIRST
+}
+```
+
+## Interesting examples {#examples}
+
+On the [TVM instructions][tvm-instructions] page, you may have noticed that the "signatures" of instructions are written in a special form called _stack notation_, which describes the state of the stack before and after the given instruction is executed.
+
+For example, `x y - z` describes an instruction that grabs two values `x` and `y` from the stack, with `y` at the top of the stack and `x` second to the top, and then pushes the result `z` onto the stack. Notice that other values deeper down the stack are not accessed.
```

Comment:
Is "grabs" here the same as "takes" or "pops"?

Reply:
Yes. @anton-trunov didn't like "pops", so I wrote "grabs", "takes" and "captures" throughout the page.

Reply:
That said, I'll add a note about this somewhere early in the page. Something along the lines of (without the continuations bit):

There's no popping of values from the TVM stack to somewhere else. There could be nested stacks of continuations, but that's a slightly different story.

The main stack of the TVM is _the_ environment, _the_ entity that Tact operates within.